File: play/game/model/reload_machine.lua
local cylinder = require("game.model.rules.cylinder")

local M = {}

function M.active_load(state)
	return state and state.pending_load or nil
end

function M.lane(state)
	if state and state.bidding and state.bidding.reload_gate then
		return "gated"
	end
	return M.active_load(state) and "loading" or "clear"
end

function M.begin_load(state, pending)
	state.pending_load = pending
end

function M.clear_active_load(state)
	state.pending_load = nil
end

function M.consume_active_load(state)
	state.pending_load = cylinder.consume_pending(state.pending_load)
	return state.pending_load
end

function M.queue_bid_load(state, pending, countdown_seconds)
	state.bidding = state.bidding or {}
	if state.pending_load and state.pending_load.source == "bid" then
		state.bidding.deferred_load = pending
		state.bidding.reload_gate = {
			countdown_seconds = countdown_seconds,
			epoch = ((state.bidding.reload_gate and state.bidding.reload_gate.epoch) or 0) + 1,
		}
		return
	end
	if not pending then
		return
	end

	state.pending_load = pending
	state.bidding.deferred_load = nil
	state.bidding.reload_gate = nil
end

function M.promote_deferred_bid_load(state)
	state.pending_load = state.bidding and state.bidding.deferred_load or nil
	if state.bidding then
		state.bidding.deferred_load = nil
		state.bidding.reload_gate = nil
	end
	return state.pending_load
end

function M.reset_bid_reload(state)
	if state.bidding then
		state.bidding.deferred_load = nil
		state.bidding.reload_gate = nil
	end
end

return M

File: convex/match/reloadMachine.ts
/**
 * Orthogonal reload-state transitions.
 *
 * The lifecycle phase and decision owner deliberately live elsewhere. In
 * particular, a bid reload does not move the lifecycle out of `bidding`.
 */
import { consumePending } from './rulesCylinder'
import {
	BID_RELOAD_COUNTDOWN_SECONDS,
	type MatchState,
	type PendingLoadState,
} from './state'

export type ReloadLane = 'clear' | 'loading' | 'gated'

export function activeLoad(state: MatchState): PendingLoadState | undefined {
	return state.reload.pending
}

export function reloadLane(state: MatchState): ReloadLane {
	if (state.reload.gate) {
		return 'gated'
	}
	return state.reload.pending ? 'loading' : 'clear'
}

export function beginLoad(state: MatchState, pending: PendingLoadState): void {
	state.reload.pending = pending
}

export function clearActiveLoad(state: MatchState): void {
	state.reload.pending = undefined
}

export function consumeActiveLoad(state: MatchState): PendingLoadState | undefined {
	state.reload.pending = consumePending(state.reload.pending)
	return state.reload.pending
}

export function queueBidLoad(state: MatchState, pending?: PendingLoadState): void {
	if (state.reload.pending?.source === 'bid') {
		state.reload.deferred = pending
		state.reload.gate = {
			countdownSeconds: BID_RELOAD_COUNTDOWN_SECONDS,
			epoch: (state.reload.gate?.epoch ?? 0) + 1,
		}
		return
	}
	if (!pending) {
		return
	}

	state.reload.pending = pending
	state.reload.deferred = undefined
	state.reload.gate = undefined
}

export function promoteDeferredBidLoad(state: MatchState): PendingLoadState | undefined {
	state.reload.pending = state.reload.deferred
	state.reload.deferred = undefined
	state.reload.gate = undefined
	return state.reload.pending
}

export function resetBidReload(state: MatchState): void {
	state.reload.deferred = undefined
	state.reload.gate = undefined
}
